### PR TITLE
change the loss calling interface

### PR DESCRIPTION
**Pull request by generall**

### CODE REVIEW

**Comment by monatis** [comment]
I updated the code where positive and negative masks are calculated to take into account identical emeddings in the case of the regular loss.

I agree with the changes in the interface, but please notice that they are actually two different mining strategies.
The mining strategy in our regular loss is seeking a single (hardest) triplet for each sample in the batch.
However, the mining strategy in the XBM loss can fetch multiple hard triplets for any sample, and this is why their scale is different as you mentioned in #171.
So what's missing there is a scaling, and it can be fixed by updating two lines in the code:

```python
pos_loss = inverse_pos_pair.sum(dim=-1) / torch.count_nonzero(
            inverse_pos_pair, dim=-1
        ) # scale pos_loss

        # minimize the similarity scores for negative pairs as a loss value
        neg_loss = neg_pair.sum(dim=-1) / torch.count_nonzero(neg_pair, dim=-1) # scale neg_loss

        # combine losses for positive and negative pairs
        loss = pos_loss + neg_loss

        # get a scalar loss value
        loss = loss.mean()

        return loss
```

I observed in some quicks tests that this one yields better optimization in the metrics.
I would like to confirm this before merging. And, we can also support multiple mining strategies by adding this as another method.

**Comment by generall** [comment]
> I updated the code where positive and negative masks are calculated to take into account identical emeddings in the case of the regular loss.


I actually don't think we need to explicitly exclude same embeddings from the triplets with `torch.eye`, cause they will form especially weak triplet which will be excluded by the hard mining anyway


> However, the mining strategy in the XBM loss can fetch multiple hard triplets for any sample, and this is why their scale is different as you mentioned in https://github.com/qdrant/quaterion/pull/171.would b

In this case, would it make sense to update our regular loss calculation? 

> And, we can also support multiple mining strategies by adding this as another method.

That would be best, especially if we match mining strategy for the xbm and regular losses, so we will be sure about scaling
